pkg/ingester/flush.go

```diff
@@ -361,64 +361,70 @@ func (i *Ingester) flushChunks(ctx context.Context, fp model.Fingerprint, labelP
 	sizePerTenant := chunkSizePerTenant.WithLabelValues(userID)
 	countPerTenant := chunksPerTenant.WithLabelValues(userID)
 
-	for j, c := range cs {
-		if err := i.closeChunk(c, chunkMtx); err != nil {
+	// Acquire lock, loop over chunks, and close.
+	chunkMtx.Lock()
+	for _, c := range cs {
+		if err := c.chunk.Close(); err != nil {
+			chunkMtx.Unlock()
 			return fmt.Errorf("chunk close for flushing: %w", err)
 		}
-
-		firstTime, lastTime := loki_util.RoundToMilliseconds(c.chunk.Bounds())
-		ch := chunk.NewChunk(
-			userID, fp, metric,
-			chunkenc.NewFacade(c.chunk, i.cfg.BlockSize, i.cfg.TargetChunkSize),
-			firstTime,
-			lastTime,
-		)
-
-		if err := i.encodeChunk(ctx, ch, c); err != nil {
-			return err
+	}
+	chunkMtx.Unlock()
+
+	// Without the lock, encode & store
+	flushTimes := make([]struct {
+		t  time.Time
+		ch chunk.Chunk
+	}, 0, len(cs))
```

Comment:
This PR kind of feels like trading CPU for memory with the side effect of more complexity/harder to maintain.  

The overhead of calling a mutex lock/unlock a few hundred or even a few thousand times feels better to me than allocating a slice over all the chunks which I would wonder has a higher garbage collection CPU overhead than the mutex calls.  I think if we want to go this route it might be better to store a pointer in this slice than copy the chunk.

It also feels a little more brittle to me now that future maintainers have to closely track this lock instead of relying on `defer` to unlock.

I guess I'm not convinced this would actually save an CPU and I think it makes the code harder to maintain

Reply:
Fair enough :+1: